### PR TITLE
Add gray-box term-discovery example

### LINKED ISSUE #1834: Example request: gray-box discovery of an unknown PDE term
# Example request: gray-box discovery of an unknown PDE term

## What I'd like to add

An example that recovers the unknown *functional form* of a term in a PDE, not
just an unknown scalar coefficient. The setup is a closure/source term whose
shape isn't known — the standard situation in subgrid/turbulence and reaction
modeling.

## Why it fits the existing design

Inverse PINNs in `physicsnemo.sym` already recover unknown scalar coefficients by
detaching known fields and letting a small network produce the unknown value. The
same graph machinery covers an unknown *term*: I declare the term as a `Function`
in the SymPy residual and provide it with a second network that maps the state to
the term value. The residual assembly (`SympyToTorch` / `Add.make_args`) and
autodiff are unchanged. So this needs no core change — it's an example that shows
a capability the framework already supports but doesn't demonstrate.

Sketch:

```python
class GrayBoxReactionDiffusion(PDE):
    def __init__(self, D=0.02):
        x, t = Symbol("x"), Symbol("t")
        u = Function("u")(x, t)
        R = Function("R")(x, t)            # supplied by a network node
        self.equations = {"reaction_diffusion": u.diff(t) - D*u.diff(x, 2) - R}

r_net = instantiate_arch(input_keys=[Key("u")], output_keys=[Key("R")], ...)
nodes = eq.make_nodes() + [u_net.make_node("u_network"), r_net.make_node("reaction_network")]
```

Physics priors (e.g. known equilibria `R(0)=R(1)=0`) go in as ordinary
`PointwiseConstraint`s on the reaction network. In the example they matter: the
data only covers part of the state range, so the term is under-determined without
them.

## Scope

I'd start with a single self-contained example under `examples/graybox_discovery/`
(data generator, training script, config, README) plus a small test. If there's
interest, a natural follow-up is a thin helper layer — a `discover_terms=` option
on `PDE.make_nodes` that generalizes the existing `freeze_terms` indexing, and a
few reusable physics-constraint classes — but I'd keep that out of the first PR.

## Prior art

Learnable-term-in-PINN with symbolic post-processing is established (e.g.
Kiyani et al. 2023, arXiv:2305.10706; UPINN, ICML 2023; AI-Aristotle, 2024;
PIKAN gray-box, 2025). The example just makes the pattern reproducible inside
PhysicsNeMo with the physics-prior constraints attached.

## Logistics

Happy to implement and maintain it, follow the style guide (Apache headers,
docstrings, tests), and sign off commits under the DCO. Opening this first to
check for overlap and get feedback on where the example should live.

### PR DESCRIPTION
**Pull request by oroikono**
# Add gray-box term-discovery example

Closes #1834

## Description

Adds `examples/graybox_discovery/`, an example that recovers an unknown term in a
PDE (a reaction/closure term) rather than an unknown scalar coefficient.

A 1D reaction-diffusion problem `u_t = D u_xx + R(u)` is set up with `R` unknown.
A field network learns `u(x, t)` and a second network learns `R(u)`, trained
jointly from point observations. The residual is declared in SymPy with `R` left
as a `Function`; a `PhysicsInformer` evaluates it each step (autodiff for the
spatial derivative, a manual `torch.autograd.grad` for the time derivative since
autodiff spatial gradients only cover x/y/z), so no change to the
residual-assembly code is needed.

The generated data uses a non-polynomial `R(u) = 3 sin(pi u)` and only observes
`u <= 0.75`, which leaves `R` under-determined near `u = 1`. Adding the known
equilibria `R(0) = R(1) = 0` as a direct penalty on the reaction network pins
the closure in the unobserved range.

## What's included

- `examples/graybox_discovery/generate_data.py` — synthetic data
- `examples/graybox_discovery/graybox_reaction_diffusion.py` — training script
- `examples/graybox_discovery/conf/config.yaml` — config
- `examples/graybox_discovery/README.md` — description, run instructions, and
  the recovered-vs-true R(u) plot
- `examples/graybox_discovery/graybox_reference.py` — numpy-only reproduction
  of the mechanism (no PhysicsNeMo/GPU required)
- `examples/graybox_discovery/test_graybox_verifiers.py` — checks the prior
  reduces the spread of the recovered closure and improves recovery in the
  unobserved region (per `CONTRIBUTING.md`, example tests live under
  `examples/`, not the top-level `test/`)

No changes to core APIs.

## Testing

`pytest examples/graybox_discovery/test_graybox_verifiers.py` runs the
reference-based checks with no GPU/torch requirement (one additional guarded
test is skipped until the `physicsnemo/sym/eq/verifiers.py` follow-up PR
lands). The example itself was run end to end in a real PhysicsNeMo
environment via `python generate_data.py && python
graybox_reaction_diffusion.py`; training converges (loss ~0.12 -> ~5e-4 over
20k steps) and the recovered R(u) tracks the true closure well past the
observed range (`u_obs = 0.75`), which is the effect the physics prior is
meant to produce -- see the plot in the example README.

## Checklist

- [x] Commits signed off (DCO)
- [x] Apache-2.0 headers on new files
- [x] Docstrings on new public functions/classes
- [x] Ran `pytest examples/graybox_discovery/test_graybox_verifiers.py`
- [x] Ran `ruff check` / `ruff format --check` (clean)
- [x] Ran `test/ci_tests/header_check.py` and
      `test/ci_tests/check_docstring_coverage.py` locally (clean)
- [x] Ran `markdownlint-cli` on changed Markdown (clean)
- [x] Updated `CHANGELOG.md`

### CODE REVIEW

**Comment by oroikono** [comment]
I’ve addressed the reported path and output-directory issues, rebased on main, and pre-commit is passing. Full NVIDIA CI appears to require external-contributor approval. Could a maintainer confirm whether this example is in scope and whether you would prefer the generated artifacts removed or the PR split into smaller pieces? I’m happy to reduce the review surface.

File: examples/graybox_discovery/graybox_reaction_diffusion.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+@hydra.main(version_base="1.3", config_path="conf", config_name="config.yaml")
+def main(cfg: DictConfig) -> None:
+    """Train the field and reaction networks jointly and plot the recovered R(u)."""
+    DistributedManager.initialize()
+    dist = DistributedManager()
+
+    log = PythonLogger(name="graybox_discovery")
+    log.file_logging()
+
+    data = np.load(cfg.data.data_file)
```
[the remark this answers is not in the document]
Comment (oroikono) [reply examples/graybox_discovery/graybox_reaction_diffusion.py]: Fixed in 028da...: the default data path is now resolved relative to the example directory.

File: examples/graybox_discovery/generate_data.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    u_data = (u[obs] + NOISE * u.std() * rng.standard_normal(obs.sum()))[:, None]
+
+    # thin the observations and sample interior collocation points
+    keep = rng.choice(len(x_data), size=min(4000, len(x_data)), replace=False)
+    x_data, t_data, u_data = x_data[keep], t_data[keep], u_data[keep]
+
+    ci = rng.choice(X.size, size=4000, replace=False)
+    x_pde, t_pde = X.ravel()[ci][:, None], Tm.ravel()[ci][:, None]
+
+    np.savez(
```
[the remark this answers is not in the document]
Comment (oroikono) [reply examples/graybox_discovery/generate_data.py:87]: Fixed in 028da...: generated data now defaults to the example directory rather than the caller’s working directory.

**Comment by oroikono** [comment]
Gentle follow-up here, tagging @ktangsali since this extends the `inverse_pinns` example you authored.

Status: the two Greptile findings (CWD-relative data/output paths) are fixed, the branch is rebased on `main`, and pre-commit is green — full CI is still gated on external-contributor approval.

Could you let me know whether a gray-box term-discovery example is in scope for `examples/`? If so, I'm glad to reduce the review surface however's easiest — e.g. drop the committed `.npz`/`.png` artifacts, or split the reference/tests out. Happy to adjust. Thanks!
